src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodTranslator.cs

```diff
@@ -73,20 +86,31 @@ public SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadO
             // since non-PG SQL does not support arrays. If the list is a constant we leave it for regular IN
             // (functionality the same but more familiar).
 
-            // TODO: The following does not work correctly if there are any nulls in a parameterized array, because
-            // of null semantics (test: Contains_on_nullable_array_produces_correct_sql).
-            // https://github.com/aspnet/EntityFrameworkCore/issues/15892 tracks caching based on parameter values,
-            // which should allow us to enable this and have correct behavior.
-
-            // We still apply this translation when it's on a column expression, since that can't work anyway with
-            // EF Core's parameter to constant expansion
+            // Note: we exclude constant array expressions from this PG-specific optimization since the general
+            // EF Core mechanism is fine for that case. After https://github.com/aspnet/EntityFrameworkCore/issues/16375
+            // is done we may not need the check any more.
+            // Note: we exclude arrays/lists over Nullable<T> since the ADO layer doesn't handle them (but will in 5.0)
 
             if (method.IsClosedFormOf(Contains) &&
-                arrayOperand is ColumnExpression &&
-                //!(arrayOperand is SqlConstantExpression) &&   // When the null semantics issue is resolved
-                _sqlExpressionFactory.FindMapping(arrayOperand.Type) != null)
+                _sqlExpressionFactory.FindMapping(operand.Type) != null &&
+                !(operand is SqlConstantExpression) &&
+                !(operandElementType.IsGenericType && operandElementType.GetGenericTypeDefinition() == typeof(Nullable<>)))
             {
-                return _sqlExpressionFactory.ArrayAnyAll(arguments[1], arrayOperand, ArrayComparisonType.Any, "=");
+                var item = arguments[1];
+                // We require a null semantics check in case the item is null and the array contains a null.
+                // Advanced parameter sniffing would help here: https://github.com/aspnet/EntityFrameworkCore/issues/17598
+                return _sqlExpressionFactory.OrElse(
+                    // We need to coalesce to false since 'x' = ANY ({'y', NULL}) returns null, not false
+                    // (and so will be null when negated too)
+                    _sqlExpressionFactory.Coalesce(
+                        _sqlExpressionFactory.ArrayAnyAll(item, operand, ArrayComparisonType.Any, "="),
+                        _sqlExpressionFactory.Constant(false)),
+                    _sqlExpressionFactory.AndAlso(
+                        _sqlExpressionFactory.IsNull(item),
+                        _sqlExpressionFactory.IsNotNull(
+                            _sqlExpressionFactory.Function("array_position", new[] {
+                                operand, _sqlExpressionFactory.Fragment("NULL")
+                            }, typeof(int)))));
```

Comment:
Looks like the last parameter relates to line 110 and not 112. Let's fix this.

Reply:
Sorry, didn't understand... Can you submit a code suggestion?

Reply:
```csharp
                        _sqlExpressionFactory.IsNotNull(
                            _sqlExpressionFactory.Function("array_position",
                                new[] { operand, _sqlExpressionFactory.Fragment("NULL") }, typeof(int)))));
// or
                        _sqlExpressionFactory.IsNotNull(
                            _sqlExpressionFactory.Function(
                                "array_position",
                                new[] { operand, _sqlExpressionFactory.Fragment("NULL") },
                                typeof(int)))));
```

Reply:
Oh OK, yeah, splitting the arguments to different lines looks better - made the change.

BTW github has a nice feature to suggest a code change on a PR - not a normal comment - which I can just accept (it becomes a new commit in the PR).